common/src/main/java/com/microsoft/identity/common/internal/authscheme/DevicePopManagerImpl.java

```diff
@@ -0,0 +1,749 @@
+import android.security.keystore.StrongBoxUnavailableException;
+import android.util.Base64;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
+
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.internal.controllers.TaskCompletedCallbackWithError;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.crypto.RSASSASigner;
+import com.nimbusds.jose.crypto.impl.RSAKeyUtils;
+import com.nimbusds.jose.jwk.RSAKey;
+import com.nimbusds.jose.util.Base64URL;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.math.BigInteger;
+import java.net.URL;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.UnrecoverableEntryException;
+import java.security.cert.CertificateException;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.RSAKeyGenParameterSpec;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.security.auth.x500.X500Principal;
+
+import static com.microsoft.identity.common.exception.ClientException.BAD_KEY_SIZE;
+import static com.microsoft.identity.common.exception.ClientException.INTERRUPTED_OPERATION;
+import static com.microsoft.identity.common.exception.ClientException.INVALID_ALG;
+import static com.microsoft.identity.common.exception.ClientException.INVALID_PROTECTION_PARAMS;
+import static com.microsoft.identity.common.exception.ClientException.JSON_CONSTRUCTION_FAILED;
+import static com.microsoft.identity.common.exception.ClientException.JWT_SIGNING_FAILURE;
+import static com.microsoft.identity.common.exception.ClientException.KEYSTORE_NOT_INITIALIZED;
+import static com.microsoft.identity.common.exception.ClientException.NO_SUCH_ALGORITHM;
+import static com.microsoft.identity.common.exception.ClientException.NO_SUCH_PROVIDER;
+import static com.microsoft.identity.common.exception.ClientException.THUMBPRINT_COMPUTATION_FAILURE;
+import static com.microsoft.identity.common.internal.net.ObjectMapper.ENCODING_SCHEME;
+
+public class DevicePopManagerImpl implements IDevicePopManager {
```

Comment:
TODO: Document/plan _exactly_ how this works with caching. Currently there is a single pop KeyPair -- we need to account for this key disappearing/token invalidation scenarios.

Reply:
Added new APIs in:
- eccf97d64d950136b4a76238996686f48a53ada7
- 8efb91e12abcd89bce72596d07c6d40afe7600c2

Current plan is:
- When generating _new_ keys, test none exist
- When _renewing_ tokens, assert that a keypair exists, send in `req_cnf`
    * Make sure correct `thumbprint` is saved as `kid` when writing tokens to cache
- When using an existing `access_token` to derive a new `SHR` for a new resource....
    * How does that work?